cpp/roaring64map.hh

```diff
@@ -1098,90 +1100,69 @@ public:
-                        uniteBytes(((iter_data *)inner_iter_data)->high_bits,
-                                   low_bits));
-                    ((iter_data *)inner_iter_data)->first_char = ',';
-                    return true;
-                },
-                (void *)&outer_iter_data);
-            std::for_each(
-                ++map_iter, roarings.cend(),
-                [&outer_iter_data](
-                    const std::pair<const uint32_t, Roaring> &map_entry) {
-                    outer_iter_data.high_bits = map_entry.first;
-                    map_entry.second.iterate(
-                        [](uint32_t low_bits, void *inner_iter_data) -> bool {
-                            ((iter_data *)inner_iter_data)->str +=
-                                ((iter_data *)inner_iter_data)->first_char;
-                            ((iter_data *)inner_iter_data)->str +=
-                                std::to_string(uniteBytes(
-                                    ((iter_data *)inner_iter_data)->high_bits,
-                                    low_bits));
-                            return true;
-                        },
-                        (void *)&outer_iter_data);
-                });
-        } else
-            outer_iter_data.str = '{';
-        outer_iter_data.str += '}';
-        return outer_iter_data.str;
+        std::string result;
+        auto sink = [&result](const char *s) {
+            result += s;
+        };
+        printToSink(sink);
+        return result;
+    }
+
+    /**
+     * Prints the contents of the bitmap to a caller-provided sink function.
+     */
+    void printToSink(const std::function<void(const char *)> &sink) const {
+        struct context_t {
+            const std::function<void(const char *)> &sink;
+            uint32_t high_bits;
+            bool needs_comma;
+        };
+        context_t context{sink, 0, false};
+
+        auto streamElement = [](uint32_t low_bits, void *arg) {
+            auto *ctx = static_cast<context_t *>(arg);
+
+            // max uint64_t is 18446744073709551615 which is 20 digits.
+            // And we need one more byte for the optional comma and one byte
+            // for the terminating \0. We provide 32 bytes here, which is ample.
+            char buffer[32];
+            char *destp = buffer;
+
+            if (ctx->needs_comma) {
+                *destp++ = ',';
+            }
+            ctx->needs_comma = true;
+
+            auto value = uniteBytes(ctx->high_bits, low_bits);
+            // -1 because we might have used a byte for the comma.
+            snprintf(destp, sizeof(buffer) - 1, "%" PRIu64, value);
+            ctx->sink(buffer);
+
+            return true;
+        };
+
+        sink("{");
+        for (const auto &entry : roarings) {
+            context.high_bits = entry.first;
+            const auto &bitmap = entry.second;
+            bitmap.iterate(streamElement, &context);
```

Comment:
I think it would be cleaner to use an iterator to iterate over the inner map, i.e. `for (uint32_t low_bits : bitmap) {`. That way there is no indirection through the lambda and we can get rid of the context struct.

Reply:
Awesome suggestion, thank you. Please take another look.